tests/interfaces/test_torch.py

```diff
@@ -690,3 +691,395 @@ def compose(f, x):
+
+        assert all([i.shape == () for i in unflattened[0]])
+        assert unflattened[1].numel() == 0
+
+    def test_model_nested_tensor(self):
+        """Test that the function correctly unflattens when the model is a nested tensor,
+        which should result in unflatten_torch returning a list of tensors of the same shape"""
+        model = [torch.ones(3), torch.ones((2, 2)), torch.ones((3, 1)), torch.ones((1, 2))]
+        unflattened = unflatten_torch(self.flat, model)
+
+        assert all(
+            [u.shape == model[i].shape for i, u in enumerate(unflattened[0])]
+        )
+        assert unflattened[1].numel() == 0
+
+
+class TestParameterHandlingIntegration:
+    """Test that the parameter handling for differentiable/non-differentiable
+    parameters works correctly."""
+
+    def test_differentiable_parameter_first(self, mocker):
+        """Test that a differentiable parameter used as the first
+        argument is correctly evaluated by QNode.jacobian, and that
+        all other non-differentiable parameters are ignored"""
+        spy = mocker.spy(qml.qnodes.JacobianQNode, "jacobian")
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="torch")
+        def circuit(weights, data1, data2):
+            # non-differentiable quantum function
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            # differentiable quantum function
+            qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
+            # non-differentiable quantum function
+            qml.templates.AngleEmbedding(data2, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        # differentiating the circuit wrt the weights
+        # input weights
+        weights = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)
+        weights = torch.tensor(weights, requires_grad=True)
+
+        # input data
+        data1 = torch.tensor([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+        data2 = torch.tensor([1, 1], requires_grad=False)
+
+        loss = circuit(weights, data1, data2)
+        loss.backward()
+
+        # check that weights is only once differentiable
+        assert weights.grad.requires_grad == False
+
+        res = weights.grad.detach().numpy()
+
+        # we do not check for correctness, just that the output
+        # is the correct shape
+        assert res.shape == weights.shape
+
+        # check that the parameter shift was only performed for the
+        # differentiable elements of `weights`, not the data input
+        num_w1 = weights.numel()
+        assert spy.call_args[1]["wrt"] == set(range(num_w1))
+
+    def test_differentiable_parameter_middle(self, mocker):
+        """Test that a differentiable parameter provided as the middle
+        argument is correctly evaluated by QNode.jacobian, and that
+        all other non-differentiable parameters are ignored"""
+        spy = mocker.spy(qml.qnodes.JacobianQNode, "jacobian")
```

Comment:
How did I not know about this before!

Reply:
Does using `mocker` require the `pytest-mock` package? What is the protocol for specifying test packages?

Reply:
Oh I see `pytest-mock` in `.travis.yml`, maybe that's enough!

Reply:
`pytest-mock` is essentially just a wrapper around `unittest.mock`, but it's a much nicer UI :)